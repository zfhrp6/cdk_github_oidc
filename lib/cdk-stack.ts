import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as iam from 'aws-cdk-lib/aws-iam';

export interface GitHubOidcStackProps extends cdk.StackProps {
  roles: {
    roleName: string;
    repositories: string[];
    isRepositoryExcactlyEquals: boolean;
    managedPolicies: string[];
    policies: {
      actions: string[];
      resources: string[];
    };
  }[];
}

export class GitHubOidcStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: GitHubOidcStackProps) {
    super(scope, id, props);

    const provider = new iam.OpenIdConnectProvider(this, 'Provider', {
      url: 'https://token.actions.githubusercontent.com',
      clientIds: ['sts.amazonaws.com'],
    });

    props.roles.forEach((params) => {
      const principalProps = params.isRepositoryExcactlyEquals
        ? {
            StringEquals: {
              'token.actions.githubusercontent.com:aud': 'sts.amazonaws.com',
              'token.actions.githubusercontent.com:sub': params.repositories,
            },
          }
        : {
            StringEquals: {
              'token.actions.githubusercontent.com:aud': 'sts.amazonaws.com',
            },
            StringLike: {
              'token.actions.githubusercontent.com:sub': [params.repositories],
            },
          };

      const role = new iam.Role(this, `github-oidc-role-${params.roleName}`, {
        roleName: params.roleName,
        assumedBy: new iam.WebIdentityPrincipal(provider.openIdConnectProviderArn, principalProps),
      });

      // attach AWS Managed Policy
      if (params.managedPolicies.length > 0) {
        params.managedPolicies.forEach((policyName) => {
          role.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName(policyName));
        });
      }

      if (params.policies.actions.length > 0) {
        console.log(params.policies.actions);
        // add other policy
        role.addToPolicy(
          new iam.PolicyStatement({
            actions: params.policies.actions,
            resources: params.policies.resources,
          })
        );
      }
    });
  }
}
