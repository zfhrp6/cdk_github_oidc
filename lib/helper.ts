export const lowerCamelCase = (s: string): string => {
  return s
    .replace(/(_[a-zA-Z])/g, (group) => group.toUpperCase())
    .replace(/[-_]/g, '')
    .replace(/^./, (group) => group.toLowerCase());
};

export const upperCamelCase = (s: string): string => {
  return lowerCamelCase(s).replace(/^./, (group) => group.toUpperCase());
};

export const snakeCase = (s: string): string => {
  return s
    .replace(/[A-Z]/g, (group) => '_' + group.toLowerCase())
    .replace(/^_/, '')
    .replace(/(__|--)/g, '_');
};
