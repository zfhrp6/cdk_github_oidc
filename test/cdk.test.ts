import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { GitHubOidcStack } from '../lib/cdk-stack';
import { testConfig } from './test-env';

test('STack snapshot test', () => {
  const app = new cdk.App();
  const stack = new GitHubOidcStack(app, 'SnapshotTestStack', testConfig);
  // スタックからテンプレート(JSON)を生成
  const template = Template.fromStack(stack).toJSON();

  // 生成したテンプレートとスナップショットが同じか検証
  expect(template).toMatchSnapshot();
});
