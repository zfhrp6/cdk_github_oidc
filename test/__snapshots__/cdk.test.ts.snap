// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`STack snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0": {
      "DependsOn": [
        "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "977fc1649d2dbcce16e23f6332faef6fa0f48aa74a0afe35f4a3467754e20cd8.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "iam:CreateOpenIDConnectProvider",
                    "iam:DeleteOpenIDConnectProvider",
                    "iam:UpdateOpenIDConnectProviderThumbprint",
                    "iam:AddClientIDToOpenIDConnectProvider",
                    "iam:RemoveClientIDFromOpenIDConnectProvider",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Provider2281708E": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ClientIDList": [
          "sts.amazonaws.com",
        ],
        "CodeHash": "977fc1649d2dbcce16e23f6332faef6fa0f48aa74a0afe35f4a3467754e20cd8",
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0",
            "Arn",
          ],
        },
        "Url": "https://token.actions.githubusercontent.com",
      },
      "Type": "Custom::AWSCDKOpenIdConnectProvider",
      "UpdateReplacePolicy": "Delete",
    },
    "githuboidcroletestrolename1policy2D2102C2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringEquals": {
                  "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
                  "token.actions.githubusercontent.com:sub": [
                    "repo:github-user/example-repository:example-oidc-branch-1",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": {
                  "Ref": "Provider2281708E",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "test-role-name-1-policy",
      },
      "Type": "AWS::IAM::Role",
    },
    "githuboidcroletestrolename1policyDefaultPolicyC883E853": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "test:TestAwsActionName",
              "Effect": "Allow",
              "Resource": "testResource",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "githuboidcroletestrolename1policyDefaultPolicyC883E853",
        "Roles": [
          {
            "Ref": "githuboidcroletestrolename1policy2D2102C2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "githuboidcroletestrolename2managedPolicyA31862DA": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringEquals": {
                  "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
                },
                "StringLike": {
                  "token.actions.githubusercontent.com:sub": [
                    [
                      "repo:github-user/example-repository:*",
                    ],
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": {
                  "Ref": "Provider2281708E",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AWSIoTEventsReadOnlyAccess",
              ],
            ],
          },
        ],
        "RoleName": "test-role-name-2-managedPolicy",
      },
      "Type": "AWS::IAM::Role",
    },
    "githuboidcroletestrolename3policyandmanagedPolicy98540753": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringEquals": {
                  "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
                },
                "StringLike": {
                  "token.actions.githubusercontent.com:sub": [
                    [
                      "repo:github-user/example-repository:*",
                    ],
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": {
                  "Ref": "Provider2281708E",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AWSIoTEventsReadOnlyAccess",
              ],
            ],
          },
        ],
        "RoleName": "test-role-name-3-policy-and-managedPolicy",
      },
      "Type": "AWS::IAM::Role",
    },
    "githuboidcroletestrolename3policyandmanagedPolicyDefaultPolicyDA5A5874": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "test:TestAwsActionName2",
              "Effect": "Allow",
              "Resource": "arn:aws:iam::123456789012:resource/test-resource-name",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "githuboidcroletestrolename3policyandmanagedPolicyDefaultPolicyDA5A5874",
        "Roles": [
          {
            "Ref": "githuboidcroletestrolename3policyandmanagedPolicy98540753",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
